Show parent groups in navbar

Martchus opened the pull request:
* Since there is currently no parent group overview
  the link just leads to the editor.
* Parentless groups are still displayed (at the top).

aaannz commented:
Is the D&D editor supposed to be part of this PR?

Martchus commented:
When starting I thought it makes most sense to continue on top of the branch for D&D editor. However, both features are independent so I remove it here.

I also noticed that I should implement tests for this.

okurz commented:
> I also noticed that I should implement tests for this.

maybe we don't need tests ;-)

okurz commented:
@Martchus you "have" tests but maybe they are not automatic. Can you maybe give a description of how you test so that we can try to reproduce that? A Of course, screenshots help, too. You don't need all that if you have automatic tests as they are "self-descriptive".

Martchus commented:
D&D editor and this are not independent (as I thought before) because this needs the editor for parent job groups (it is shown when you click on a parent).

I have also a variant of this introducing a parent group overview (with accumulated build results) which is shown when clicking on a parent but it is still WIP: https://github.com/Martchus/openQA/tree/parent_group_overview

So I'd propose D&D editor should be merged first. Then I implement basic PhantomJS tests for this but test the submenu being shown on hover manually.

Because you asked, that is how it looks like:
![screenshot](https://cloud.githubusercontent.com/assets/10248953/19923350/3bf89c64-a0e7-11e6-8d4d-283cf7d1af70.png)
Steps to tests are so obvious that I thought it is not worth mentioning. I mean, just open the menu. Of course you need some parents (either use D&D editor or API routes for that).

aaannz commented:
Is this ready to be merged or are you still working on tests? LGTM otherwise.

okurz commented:
I guess somehow I forgot to submit inline comments or they got lost. But issues I see:

* no tests (already discussed, ok for now)
* parent groups should be at top of list, makes loosing the menu focus on scrolling less likely
* adding a new file under MIT license but not supplying the MIT license, I don't know if this is an issue
* the bootstrap bootmenu code style is … different, e.g. some magic numbers

Martchus commented:
> parent groups should be at top of list, makes loosing the menu focus on scrolling less likely

Fixed

> adding a new file under MIT license but not supplying the MIT license, I don't know if this is an issue

Don't know, too.

> the bootstrap bootmenu code style is … different, e.g. some magic numbers

There are comments for those numbers. (JavaScript has no enums anyways so I'd say it is ok.)

aaannz commented:
I think we _should_ include MIT licence, but in a way it's not confusing with our GPL license. Something like described in https://softwareengineering.stackexchange.com/questions/234511/what-is-the-best-practice-for-arranging-third-party-library-licenses-paperwork - aka name that file LICENSE-3RD-PARTY.txt.

@lnussel @sysrich Do you have any input on licensing?

coolo commented:
is there any reason to put a copy of this file in our git? it looks downloaded

Martchus commented:
> it looks downloaded 
I don't think it is worth implementing such features myself if there's already a good solution there. We also use other 3rd party libs so I thought this is no problem.

> is there any reason to put a copy of this file in our git?
I had to change the script a little so I thought it is best to include it directly in the project. The MIT license should allow it. Additionally, it might not be a good idea to rely on external sources for every little helper script we include.

However, I'm also not sure how to include the license info best. The approach from the page @aaannz linked would make sense in my opinion.

Martchus commented:
I now added a file `COPYING-Bootstrap-submenu`. The name should already say that this file only applies to 'Bootstrap-submenu' and I also added a list of the particular files.

I also extended the comment in the file I modified to make it more clear that this is an edited version.

aaannz commented on COPYING-Bootstrap-submenu:
```diff
@@ -0,0 +1,24 @@
+- assets/css/bootstrap-submenu.css
+
+Copyright Vasily A., 2014-2016
+Copyright openQA Team, 2016
```
This should be SUSE LLC

coolo commented:
How exactly did you have to change it? I still don't like including copies - especially if they are changed. In the future you might need to rebase your changes to the latest version of it.

coolo commented:
So you changed 2 lines and for that added SUSE's copyright?

okurz commented on templates/layouts/bootstrap.html.ep:
```diff
@@ -61,13 +61,30 @@
+                                            </ul>
+                                        % }
+                                    </li>
+                                  % }
```
quite some indendation in this block but ok for me in this case

coolo replied:
What I worry about is the level of DB queries done in the layout. This should be stashed somewhere globally. E.g. in a before_render hook.

Martchus replied:
I also noticed this but I didn't want to mix refactoring with this commit. But I can improve this afterwards in the way you propose.

BTW: Database access is done in the view anyways, just in a more hidden way.

okurz replied:
oh refactoring is no problem. remember TDD (1. failing test, 2. fix, 3. refactor) and of course you can do refactoring before you add new stuff, all covered by tests :-)

coolo commented:
So I tested with the original js and couldn't find any problem, so I'm objecting the inclusion in our git.

Martchus commented:
There is no problem. I just altered the behaviour to show sub menus on hover (instead of on click). Then the parent group overview can be accessed on click. 

I could also fork it on GitHub, do the changes there and just link the file here.

coolo commented:
Or you offer the upstream author an option, create a PR to the upstream repo and link the file there.

coolo commented:
where there means your fork while waiting for the PR to be merged.

coolo commented:
But the 'parent group overview' currently links to a page only admins can access. So there is little point on your feature for now.

Martchus commented:
As soon as the parent group overview is merged this feature becomes relevant. So I'll fork it, make the change optional and maybe the PR gets accepted.

Martchus commented:
This is using the version from my fork. Making the hover thing optional (just to make the change for this tiny script go upstream) seems a little bit over-engineering. Additionally, this change must be done in an appropriate way to go upstream - however my knowledge about the design/structure of this project is limited. So from my point of view it is not worth it. Just using the version with 2 lines altered is just so much simpler and effort for re-basing should be minimal if even required.

coolo commented:
at least file a feature request linking to your change.

Martchus commented:
There is already an issue (with a proposal for a change): https://github.com/vsn4ik/bootstrap-submenu/issues/13

I guess everyone who wants that just changes it himself for now :-/
